Fog move to Pixel shader

medievil1 opened the pull request:
3rd times the charm, accidentally kept a commit that had the vertex passthru/program shader changes, this time it is clean just fog changes

medievil1 committed:
fog stuff

medievil1 committed:
add FF pixelshader fog

medievil1 committed:
fix formatting issues

address review comment

medievil1 committed:
crash fix....

everything in this branch crashes on startup unless there is some kind of delay here
Sleep(0) also works
key thing is the routine seems to try to run before everything is set

medievil1 committed:
just adding more info to comment

medievil1 committed:
address review comment

changed fog enable/ fog table code flow

PatrickvL commented on src/core/hle/D3D8/XbPixelShader.cpp:
```diff
@@ -1485,6 +1512,17 @@ void DxbxUpdateActivePixelShader() // NOPATCH
 	  }
 	  fColor[PSH_XBOX_CONSTANT_FRONTFACE_FACTOR].r = frontfaceFactor;
 
+	  fogEnable = XboxRenderStates.GetXboxRenderState(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGENABLE) > 0;
```
This reads a float, does a check turning it into a float and then assigns the result to a float - but it's not very clear what float value that is.

Perhaps it's good to make that more explicit somehow, perhaps using a ternary expression setting between two constants (FLOAT_TRUE vs FLOAT_FALSE for instance)?

medievil1 replied:
changed to  
`fogEnable = (NV2ARenderStates.GetXboxRenderState(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGENABLE) > 0 == Fog_Enabled) ? Fog_Enabled : Fog_Disabled;` and 
`fogEnable = (XboxRenderStates.GetXboxRenderState(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGENABLE) > 0 == Fog_Enabled) ? Fog_Enabled : Fog_Disabled;`
 and added  
 float Fog_Enabled = 1;
 float Fog_Disabled = 0;
 Verified it does work

jackchentwkh replied:
GetXboxRenderState() returns a DWORD not a float. Patrick might mistake it as GetXboxRenderStateAsFloat(). So the original code should work correctly. but it would be more feasible to use 
  fogEnable = XboxRenderStates.GetXboxRenderState(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGENABLE) != 0;
I would suggest to use the code above or 
  fogEnable = XboxRenderStates.GetXboxRenderState(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGENABLE) != false;

PatrickvL commented on src/core/hle/D3D8/XbPixelShader.cpp:
```diff
@@ -1284,7 +1302,7 @@ void DxbxUpdateActivePixelShader() // NOPATCH
   D3DXCOLOR fColor[PSH_XBOX_CONSTANT_MAX];
   extern XboxTextureStateConverter NV2ATextureStates;
-  
+  float fogDensity, fogStart, fogEnd, fogTableMode, fogEnable;
```
Just a small nitpick: Consider ordering these variables in the same order as they're being referred-to in the code?

medievil1 replied:
fixed

PatrickvL commented on src/core/hle/D3D8/XbPixelShader.cpp:
```diff
@@ -1179,6 +1190,13 @@ void UpdateFixedFunctionPixelShaderState()
+		ffPsState.FogStart = XboxRenderStates.GetXboxRenderStateAsFloat(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGSTART);
+		ffPsState.FogEnd = XboxRenderStates.GetXboxRenderStateAsFloat(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGEND);
+		
+		
```
Why more than 1 empty separator line?

medievil1 replied:
fixed (must have been a copy/paste issue)

PatrickvL commented on src/core/hle/D3D8/XbPixelShader.cpp:
```diff
@@ -672,6 +672,7 @@ PSH_RECOMPILED_SHADER CxbxRecompilePixelShader(CxbxPSDef &CompletePSDef)
 			}
 		}
 		pShader->Release();
+		
```
Needless empty separator line

medievil1 replied:
fixed

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl:
```diff
@@ -289,19 +289,9 @@ float DoFog(const VS_INPUT xIn)
-        fogFactor = (state.Fog.End - fogDepth) / (state.Fog.End - state.Fog.Start); // (end - d) / (end - start)
-
-    return fogFactor;
+    
```
Needless empty separator lines

medievil1 replied:
fixed

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionPixelShader.hlsli:
```diff
@@ -63,10 +63,16 @@ namespace FixedFunctionPixelShader {
-	const int SAMPLE_2D = 1;
-	const int SAMPLE_3D = 2;
-	const int SAMPLE_CUBE = 3;
+    const int SAMPLE_NONE = 0;
```
Indentation is different from surrounding lines

medievil1 replied:
not seeing this, it is 4 space indent above and below this line and everything looks normal in VS (white space shows the same), originally could have been tab I suppose..I will tab it and upload

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionPixelShader.hlsl:
```diff
@@ -245,6 +245,24 @@ TextureArgs ExecuteTextureStage(
 
+// Calculate the fog factor
+	// Some of this might be better done in the pixel shader?
+		float fogFactor;
```
Incorrect indentation

medievil1 replied:
fixed

PatrickvL replied:
I still see space vs tab differences compared to surrounding lines?

PatrickvL commented on src/core/hle/D3D8/Direct3D9/Direct3D9.cpp:
```diff
@@ -3285,6 +3285,7 @@ LTCG_DECL xbox::dword_xt* CxbxrImpl_MakeSpace(void)
 // which is normally the entry point for all Xbox D3D 'things'.
 void HLE_PushInit()
 {
+	SwitchToThread();
```
Please add a comment explaining succinctly (but accurately!) why this is apparently needed (and what alternatives exist but are not suitable).

Without such a comment there's bound to be confusion and high risk of removal later on

medievil1 replied:
fixed, hopefully it is clear

PatrickvL commented on src/core/hle/D3D8/Direct3D9/CxbxPixelShaderTemplate.hlsl:
```diff
@@ -1,3 +1,5 @@
+
+
```
Needless empty separator lines

PatrickvL commented on src/core/hle/D3D8/Direct3D9/CxbxPixelShaderTemplate.hlsl:
```diff
@@ -91,6 +94,8 @@ uniform const float  FRONTFACE_FACTOR : register(c27); // Note : PSH_XBOX_CONSTA
 
 // DEFINES INSERTION MARKER
 
+
```
Needless empty separator lines

medievil1 replied:
fixed

PatrickvL commented on src/core/hle/D3D8/Direct3D9/CxbxPixelShaderTemplate.hlsl:
```diff
@@ -334,6 +340,34 @@ float3 DoBumpEnv(const float4 TexCoord, const float4 BumpEnvMat, const float4 sr
+	const int FOG_TABLE_EXP2    = 2;
+	const int FOG_TABLE_LINEAR  = 3;
+
+          float fogFactor;
```
Incorrect indentation

medievil1 replied:
fixed

PatrickvL reviewed (approved):
Some non-functional suggestions, otherwise good

medievil1 committed:
address review comment

medievil1 committed:
address review comments

FF templates, for some reason didn't take first time, comments should be fixed now

PatrickvL commented on src/core/hle/D3D8/XbPixelShader.cpp:
```diff
@@ -1284,7 +1298,9 @@ void DxbxUpdateActivePixelShader() // NOPATCH
   extern XboxTextureStateConverter NV2ATextureStates;
-  
+  float fogEnable, fogTableMode, fogDensity, fogStart, fogEnd;
+  float Fog_Enabled = 1;
```
These "enabled" and "disabled" variables can be marked 'const' and could apply to many other "float booleans", so would ideally be moved to a more central location and used everywhere appropriate

PatrickvL replied:
If you are going to rename them, consider `FloatTrue` vs `FloatFalse` or something along those lines.

PatrickvL commented on src/core/hle/D3D8/XbPixelShader.cpp:
```diff
@@ -1395,7 +1411,17 @@ void DxbxUpdateActivePixelShader() // NOPATCH
 	  }
 	  fColor[PSH_XBOX_CONSTANT_FRONTFACE_FACTOR].r = frontfaceFactor;
-
+	  fogEnable = (NV2ARenderStates.GetXboxRenderState(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGENABLE) > 0 == Fog_Enabled) ? Fog_Enabled : Fog_Disabled;
```
Please remove the ` == Fog_Enabled` part, because its inadvertently inverting the result while the "comparison above zero" already results in a boolean. (Turning that boolean into one of the "enabled" or "disabled" constants in the rest of the ternary expression is fine, besides their name and declaration which I commented on elsewhere.)

jackchentwkh replied:
@PatrickvL the GetXboxRenderState() returns a DWORD instead of float. original code should work correctly. 
fogEnable = NV2ARenderStates.GetXboxRenderState(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGENABLE) !=false;
or even directly assign the renderstate to fogEnable should work.

PatrickvL replied:
Ah, that I overlooked, probably because all assignments are targeting float variants, so apparently those are all doing an implicit (instead of an explicit) type-cast.

Still, my remark would make it somewhat more clear that this specific variable contains a boolean. What do you think is wisest to do here? Be more explicit or rely on implicit behaviour?

(Also, remember that later on when the port to Direct3D 11 takes shape, shader variables can become integral or boolean besides the current float-only type.)

jackchentwkh replied:
how about this 
const FloatTrue =1.0
const FloatFalse = 0;
fogEnable = NV2ARenderStates.GetXboxRenderState(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGENABLE) !=false?FloatTrue:FloatFalse;

PatrickvL replied:
That can work, sure.
Those two consts would be good to have indeed.

Except `X != false` is a double negative
While `X > 0` is more easy on my eyes.

So medievil, you choose!

medievil1 replied:
x > 0 it is...
both are readable and understandable, chose it because it is less characters/space....lol

I placed the FloatTrue/False inside Direct3d9.h so it can be accessed by all of the graphics code

PatrickvL commented on src/core/hle/D3D8/XbPixelShader.cpp:
```diff
@@ -1485,6 +1511,17 @@ void DxbxUpdateActivePixelShader() // NOPATCH
 	  }
 	  fColor[PSH_XBOX_CONSTANT_FRONTFACE_FACTOR].r = frontfaceFactor;
 
+	  fogEnable = (XboxRenderStates.GetXboxRenderState(xbox::_X_D3DRENDERSTATETYPE::X_D3DRS_FOGENABLE) > 0 == Fog_Enabled) ? Fog_Enabled : Fog_Disabled;
```
Same comments as above applies here as well

PatrickvL commented on src/core/hle/D3D8/Direct3D9/CxbxPixelShaderTemplate.hlsl:
```diff
@@ -334,6 +336,34 @@ float3 DoBumpEnv(const float4 TexCoord, const float4 BumpEnvMat, const float4 sr
 
 PS_OUTPUT main(const PS_INPUT xIn)
 {
+	//Fog
```
Doesn't the exact same code occur in the fixed function shader hlsl too? Consider moving all shared code into a single hlsli include file

medievil1 replied:
FF already has it moved to the hlsli file, however I was not sure I could include that file in the normal shader template since it may reset some variables based on FF state, I could remove it from the FF hlsli and create a new one just for fog to include in both templates, but it seemed a waste to include a file with 4 lines....

PatrickvL replied:
Agreed. Maybe later, when more symbols get to be shared between shader paths

medievil1 committed:
Address Review comments

per discussion, created FloatFalse and FloatTrue, placed them more centrally in Direct3D9.h and adjusted fogEnable  variable setter... also corrected spacing on one line in the FF pixel shader template
